src/Framework/PaymentGateways/PaymentGateway.php

```diff
@@ -76,6 +78,21 @@ public function __construct(SubscriptionModule $subscriptionModule = null)
         $this->subscriptionModule = $subscriptionModule;
     }
 
+    /**
+     * @inheritDoc
+     *
+     * @unreleased If the return type from the refundDonation() method in the child gateway cass isn't explicitly declared,
+     *             by default, it returns 'false' to prevent:
+     *             1) Duplication of refund checkboxes on gateways that already implement it;
+     *             2) Adding the refund checkbox on the admin screen in gateways with an empty refundDonation() method;
+     *             3) Overwriting legacy refund methods with empty refundDonation() methods.
+     *             4) Break gateways that implement the refundDonation() method but don't return the PaymentRefunded command as result
+     */
+    public function supportsRefund(): bool
+    {
+        return $this->isFunctionReturningCommand('refundDonation', (new PaymentRefunded()));
+    }
```

Comment:
@glaubersilva thanks for really thinking through this! I love how you outlined all the scenarios - it's really helpful for me to understand what's going on.

I do wonder if we're being a little too strict on the implementation of a return type though, as some folks might be building gateways and not be used to doing that sort of thing in php. 

For that reason, (if possible) i'm thinking we should just add a `refundDonation()` to the main PaymentGateway class that just returns false and let `supportsRefund()` use `$this->isFunctionImplementedInGatewayClass('refundDonation')`.  I like the simplicity of saying that a gateway supports refunds if it's using the refundDonation method.

To prevent duplication maybe we can use some kind of combination of `doing_filter` and/or `doing_action` before proceeding in the legacy adapter.  Another options is in our legacy adapter in the js script we just check for the checkbox in the dom to make sure it doesnt' exist yet.

Open to your thoughts here.

Reply:
Thanks for the insights! I tried an approach with the `doing_action` but it seems not to be the most suitable for this situation, so I implemented a JS approach here: 1e486288916efdb506bee0d57655c257a44ad4d2

Basically what the changes implemented in the commit mentioned above are doing is: 

1. Ensure that the framework is the last one to register a callback in the `give_view_donation_details_totals_after` hook.
2. Before rendering the checkbox on the hook from the previous step, verify if there is just one hidden checkbox in the DOM's proper place, if there is more than one, means that the gateway already added it, so the framework will do nothing.